File: Assets/src/model/grasses/Redcap.cs
using System;
using UnityEngine;

[Serializable]
[ObjectInfo("redcap", flavorText: "", description: "You may pop the Redcap, applying the Vulnerable Status to adjacent enemies for 7 turns.\nVulnerable creatures take +1 attack damage.")]
public class Redcap : Grass {
  public Redcap(Vector2Int pos) : base(pos) { }

  public void Pop(Player who) {
    OnNoteworthyAction();
    foreach (var p in floor.AdjacentActors(pos)) {
      if (p.faction == Faction.Enemy) {
        p.statuses.Add(new VulnerableStatus(7));
      }
    }
    Kill(who);
  }
}

File: web/src/model/grasses/Redcap.ts
import { Grass } from './Grass';
import { Faction, ON_TOP_ACTION_HANDLER, type IOnTopActionHandler } from '../../core/types';
import { Ground } from '../Tile';
import { VulnerableStatus } from '../statuses/VulnerableStatus';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * You may pop the Redcap, applying Vulnerable (7 turns) to adjacent enemies.
 * Port of C# Redcap.cs.
 */
export class Redcap extends Grass implements IOnTopActionHandler {
  readonly [ON_TOP_ACTION_HANDLER] = true as const;
  readonly onTopActionName = 'Pop';

  constructor(pos: Vector2Int) {
    super(pos);
  }

  static canOccupy(tile: Tile): boolean {
    return tile instanceof Ground;
  }

  handleOnTopAction(): void {
    this.pop(GameModelRef.main.player);
  }

  pop(who: any): void {
    this.onNoteworthyAction();
    for (const actor of this.floor!.adjacentActors(this.pos)) {
      if ((actor as any).faction === Faction.Enemy) {
        (actor as any).statuses.add(new VulnerableStatus(7));
      }
    }
    this.kill(who);
  }
}

entityRegistry.register('Redcap', Redcap);
